flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/dataview/MapView.java

```diff
@@ -19,94 +19,104 @@
 package org.apache.flink.table.api.dataview;
 
 import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.common.state.MapState;
 import org.apache.flink.api.common.typeinfo.TypeInfo;
 import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.table.annotation.DataTypeHint;
+import org.apache.flink.table.api.DataTypes;
 import org.apache.flink.table.dataview.MapViewTypeInfoFactory;
+import org.apache.flink.table.functions.AggregateFunction;
+import org.apache.flink.table.functions.TableAggregateFunction;
+import org.apache.flink.table.types.DataType;
 
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
+import java.util.Objects;
 
 /**
- * A {@link MapView} provides Map functionality for accumulators used by user-defined aggregate
- * functions [[AggregateFunction]].
+ * A {@link DataView} that provides {@link Map}-like functionality in the accumulator of an {@link AggregateFunction}
+ * or {@link TableAggregateFunction} when large amounts of data are expected.
  *
- * <p>A {@link MapView} can be backed by a Java HashMap or a state backend, depending on the
- * context in which the aggregation function is used.
+ * <p>A {@link MapView} can be backed by a Java {@link HashMap} or can leverage Flink's state backends
+ * depending on the context in which the aggregate function is used. In many unbounded data scenarios,
+ * the {@link MapView} delegates all calls to a {@link MapState} instead of the {@link HashMap}.
  *
- * <p>At runtime {@link MapView} will be replaced by a state MapView which is backed by a
- * {@link org.apache.flink.api.common.state.MapState} instead of {@link HashMap} if it works
- * in streaming.
+ * <p>Note: Keys of a {@link MapView} must not be null. Nulls in values are supported. For heap-based
```

Comment:
Is this correct that `Note: Keys of a {@link MapView} must not be null` Why do we need the `withNullableKeys` in `MapViewSpec` then?

Reply:
The `DISTINCT` uses `MapView` internally (see `AggregateUtil#extractDistinctInformation`). For this, we need to support nullable keys. But this is not exposed to the user. Otherwise we would need a similar logic for the `ListView`.